### PR TITLE
Handle file: dependencies

### PR DESCRIPTION
**Pull request by balsoft**
Ligo's website uses a `file:` dependency syntax in `package.json`. We need to handle that. It turned out to be pretty simple really.

### CODE REVIEW

**Comment by zhenyavinogradov** [comment]
Why does the test fail?

**Comment by balsoft** [comment]
Because I'm using `builtins.filterSource`, and I probably shouldn't. I'll try to figure it out on Monday.

**Comment by balsoft** [comment]
@zhenyavinogradov please review again, tests work now

File: default.nix [hunk 1]
```diff
@@ -117,44 +124,53 @@ with stdenv.lib; let
...
-        cp ${packageLockJson} ./package-lock.json
-
         echo 'building npm cache'
         chmod u+w ./package-lock.json
         NODE_PATH=${npmModules} node ${./mknpmcache.js} ${cacheInput "npm-cache-input.json" lock}
 
         echo 'building node_modules'
         npm $npmFlags ci
         patchShebangs ./node_modules/
+        ${concatMapStringsSep "\n" (dep: "cp -r ${dep}/node_modules/* node_modules") peerDependencies}
```
Comment (zhenyavinogradov) [inline default.nix:170]: I'm not sure I understand the logic for doing this. AFAIU this adds dependencies of local packages to `node_modules`, but I think that we should respect the behavior of `npm install`, and if it does not include these dependencies, we should not include them either

Comment (balsoft) [reply default.nix:170]: But if we don't include those dependencies, the resulting package won't work.

This is a good question: should we mimic the behaviour of `npm` even if it means more work needs to be done manually, or do we go for "least effort" approach (as I have done here)?

Comment (zhenyavinogradov) [reply default.nix:170]: Does that mean the package work if it's built with pure `npm`? I think we don't have to support a case like this then

File: default.nix [hunk 2]
```diff
@@ -90,6 +95,8 @@ with stdenv.lib; let
 
   commonBuildInputs = [ _nodejs makeWrapper ];  # TODO: git?
 
+  normalizeName = builtins.replaceStrings [ "@" "." "/" ] [ "_" "_" "_" ];
```
Comment (kirelagin) [inline default.nix:98]: Would be nice to have some documentation (or comment) for this. Or maybe a library function.

Anyway, this set of symbols looks somewhat arbitrary to me. According to [this](https://github.com/NixOS/nix/blob/a7540294cfae82c098e8691cd5212a9184add574/nix-rust/src/store/path.rs#L109) you might need to escape way more chars than this.

File: default.nix [hunk 3]
```diff
@@ -117,44 +124,53 @@ with stdenv.lib; let
...
         NODE_PATH=${npmModules} node ${./mknpmcache.js} ${cacheInput "npm-cache-input.json" lock}
 
         echo 'building node_modules'
         npm $npmFlags ci
         patchShebangs ./node_modules/
+        ${concatMapStringsSep "\n" (dep: "cp -r ${dep}/node_modules/* node_modules") peerDependencies}
 
         mkdir $out
-        mv ./node_modules $out/
+        cp -r ./node_modules $out/
```
Comment (kirelagin) [inline default.nix:173]: Why?

File: default.nix (hunk 1 above)
Comment (balsoft) [reply default.nix:170]: Yes, it will require additional commands if built with pure `npm` (`npm install` complains about peer dependencies and tells you how to install them manually). I don't know what's the reasoning for this. If there is a compelling reason not to automatically resolve peer dependencies this way, I'd be happy to remove this logic. However, if this npm behaviour is for historic reasons only, I think we should keep it the way it currently is becuase this is simply more user-friendly and simplifies the nix expressions needed to build the package.

File: default.nix (hunk 2 above)
Comment (balsoft) [reply default.nix:98]: Those are the characters that can appear in javascript package name, but can't appear in nix path name.

File: default.nix (hunk 3 above)
Comment (balsoft) [reply default.nix:173]: Because `node_modules` can contain read-only peer dependencies now, and in that case it can't be moved.

File: default.nix (hunk 1 above)
Comment (kirelagin) [reply default.nix:170]: That’s not for historic reasons, it’s actually the opposite, npm < 3 used to install peer dependencies automatically, now it doesn’t. See npm/npm#6565.

What I’m somewhat confused about is that _peer_ dependencies seem to be some special weird thing meant for plugins or something like this (I have no idea). These are not the same as local (`file:`) dependencies, are they?

Comment (balsoft) [reply default.nix:170]: Peer dependencies are transitive dependencies of local dependencies IIUC.

Comment (kirelagin) [reply default.nix:170]: Reference on peer dependencies that gets me confused: https://nodejs.org/en/blog/npm/peer-dependencies/

Comment (balsoft) [reply default.nix:170]: Oh, ok, so I need to RTFM before writing code as it turns out.

So the peer dependency change doesn't have to do anything with the main premise of this PR -- handling `file:` dependencies. It just so happens that one of the `local` dependencies of LIGO's website has peer dependencies that aren't a dependency of the website itself.

Whether my logic of automatically installing peer deps is correct or not is a complex question, because it appears that the issues you've mentioned is to do with `npm link` and other development-time commands that don't interest us much because `nix-npm-buildpackage` is targeting CIs.

**Comment by zhenyavinogradov** [comment]
Can you please also add a simple test case for this, so we can be sure that we don't break it in the future?

File: default.nix (hunk 2 above)
Comment (kirelagin) [reply default.nix:98]: Still would be nice to have some documentation.

I found [this](https://docs.npmjs.com/files/package.json) and it says that the package name can only contain “url-safe” characters. I’m not sure what this means, but I suppose, they refer to the “unreserved characters” from the URL RFC, namely:

```text
unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"
```

which means that the only problematic one is `~` (which you did not include in your list). (Note that _both_ Nix and npm allow `.`, but not as the first char.) `@` and `/` in your list, I suppose, come from the scoped packages that have the form `@<scope>/<package>` and, according to the [doc](https://docs.npmjs.com/misc/scope) the restrictions for scope names are the same.

tldr; basically, it looks like you need to replace your list with just `~` and remove leading `@`.

Comment (kirelagin) [reply default.nix:98]: (Also, tbh, I would probably replace both `~` and `/` with `-` rather than `-` for readability reasons.)

File: default.nix (hunk 1 above)
Comment (kirelagin) [reply default.nix:170]: So, IIUC, in this specific case the best solution would be to revert the peer dependency logic you added, and then... um... fix the build by either fixing the dependency that needs a peer for some reason, or adding the peer dependency to `package.yaml` of the repo?

Comment (kirelagin) [reply default.nix:170]: (I’m really with Zhenya on not doing what npm doesn’t do. I don’t really understand peer dependencies, but it kinda sounds like one’s project should declare peer dependencies of its dependencies explicitly?)

File: default.nix [hunk 4]
```diff
@@ -117,34 +124,41 @@ with stdenv.lib; let
...
-            filter = path: type: elem (getRelativePath path) usedPaths;
+            filter = path: type: any (x: elem x usedPathsRec) (dirOfRec (getRelativePath path));
           };
         in if canCleanSource src then cleanedSource else src;
 
-      packageJson = filteredSrc + "/package.json";
-      packageLockJson = filteredSrc + "/package-lock.json";
+
+      packageJson = src + "/package.json";
+      packageLockJson = src + "/package-lock.json";
```
Comment (zhenyavinogradov) [inline default.nix:146]: This path will change every time anything in `src` changes, causing everything to rebuild. As I understand the problem is that now `filteredSrc` depends on the contents of `package-lock.json`, causing infinite recursion. In that case we can create another filter expression which contains only `package.json` and `package-lock.json`, read `package-lock.json` from there, and then use it to construct a filter which will include local dependencies

Comment (balsoft) [reply default.nix:146]: Oh, good point.

Comment (kirelagin) [reply default.nix:146]: @balsoft Do you plan to address this in this PR or would you like to create a separate issue?

Comment (balsoft) [reply default.nix:146]: I think it's a separate issue.

Comment (zhenyavinogradov) [reply default.nix:146]: Why not do it in this PR?

Comment (balsoft) [reply default.nix]: Maybe fixed it here, please take a look

File: default.nix [hunk 5]
```diff
@@ -118,30 +125,46 @@ with stdenv.lib; let
...
+            filter = path: type: any (x: elem x usedPathsRec) (dirOfRec (getRelativePath path));
           };
         in if canCleanSource src then cleanedSource else src;
 
-      packageJson = filteredSrc + "/package.json";
-      packageLockJson = filteredSrc + "/package-lock.json";
+      metaSrc = cleanSourceWith {
+        src = src;
+        filter = path: type: elem (getRelativePath path) [ "package.json" "package-lock.json" ];
+      };
```
Comment (zhenyavinogradov) [inline default.nix:150]: `filteredSrc` also has a check for `canCleanSource` to support the situation when `src` is a nix store path not produced by `cleanSourceWith`, it should be done here as well. Also, to avoid repetition, you can make `filteredSrc` a function which accepts `localDeps` as a parameter and use `filteredSrc []` for `metaSrc`

File: default.nix [hunk 6]
```diff
@@ -118,30 +125,46 @@ with stdenv.lib; let
...
-          getRelativePath = path: removePrefix (toString origSrc + "/") path;
-          usedPaths = [ "package.json" "package-lock.json" ];
+          usedPaths = [ "package.json" "package-lock.json" ]
+            ++ localDeps;
+          dirOfRec = x: if dirOf x == "." || dirOf x == "/" then [x] else ([x] ++ dirOfRec (dirOf x));
+          usedPathsRec = builtins.concatMap dirOfRec usedPaths;
           cleanedSource = cleanSourceWith {
             src = src;
-            filter = path: type: elem (getRelativePath path) usedPaths;
+            filter = path: type: any (x: elem x usedPathsRec) (dirOfRec (getRelativePath path));
```
Comment (zhenyavinogradov) [inline default.nix:143]: I just noticed this, but I think `any` should not be used here, because it will include all parent directories of a dependency, while we only want the exact directory of the dependency to be included. For example, if we have dependency `file:./deps/foo`, and there also exists `./deps/extra-stuff` directory which is not used as a dependency, it will still be included in the filtered source

File: default.nix (hunk 4 above)
Comment (kirelagin) [reply default.nix]: > Why not do it in this PR?

_Because it has the potential for postponing the merging of this PR forever._

Comment (zhenyavinogradov) [reply default.nix]: In my opinion caching node_modules is more important than `file:` dependency support. @balsoft I can help you with finishing this PR if you want

Comment (zhenyavinogradov) [reply default.nix]: There're only minor fixes left as I see

**Comment by angerman** [comment]
Just ran into this issue; are we still planning on merging this feature?
